File: src/providers.ts
// Owned OAuth refresh. The gateway refreshes provider tokens itself — no
// dependency on pi's code in the credential path. Each provider needs only a
// few public constants (token endpoint, client id, body encoding) plus a
// standard `refresh_token` grant. The constants were lifted once from each
// provider's own published OAuth client; they're public, not secret.
// See docs/injection-spec.md.

export interface ProviderConfig {
  tokenUrl: string;
  clientId: string;
  encoding: "form" | "json";
  /** Subtracted from the computed expiry — matches what each provider's own
   * client bakes in (refresh a little before the real expiry). */
  skewMs: number;
}

export const PROVIDERS: Record<string, ProviderConfig> = {
  "openai-codex": {
    tokenUrl: "https://auth.openai.com/oauth/token",
    clientId: "app_EMoamEEZ73f0CkXaXp7hrann",
    encoding: "form",
    skewMs: 0,
  },
  anthropic: {
    tokenUrl: "https://platform.claude.com/v1/oauth/token",
    clientId: "9d1c250a-e61b-44d9-88ed-5944d1962f5e",
    encoding: "json",
    skewMs: 5 * 60 * 1000,
  },
};

export interface RefreshedTokens {
  access: string;
  refresh: string;
  expires: number;
}

/** Map a provider token-endpoint response to our credential fields. Pure
 * (aside from Date.now for the absolute expiry). Throws on a malformed
 * response so the caller fails closed. */
export function mapTokenResponse(cfg: ProviderConfig, json: unknown): RefreshedTokens {
  const j = json as { access_token?: unknown; refresh_token?: unknown; expires_in?: unknown };
  if (typeof j?.access_token !== "string" || typeof j?.refresh_token !== "string" || typeof j?.expires_in !== "number") {
    throw new Error(`token response missing fields: ${JSON.stringify(json)}`);
  }
  return {
    access: j.access_token,
    refresh: j.refresh_token,
    expires: Date.now() + j.expires_in * 1000 - cfg.skewMs,
  };
}

/** Perform the OAuth `refresh_token` grant for a provider. A direct host-side
 * call (the gateway's own trusted action, not the box's egress). Throws on any
 * failure — the caller must fail closed, never inject a stale token. */
export async function refreshOAuth(name: string, refreshToken: string): Promise<RefreshedTokens> {
  const cfg = PROVIDERS[name];
  if (!cfg) throw new Error(`no refresh config for provider "${name}"`);
  const params = { grant_type: "refresh_token", refresh_token: refreshToken, client_id: cfg.clientId };
  const [body, contentType] =
    cfg.encoding === "form"
      ? [new URLSearchParams(params).toString(), "application/x-www-form-urlencoded"]
      : [JSON.stringify(params), "application/json"];

  let res: Response;
  try {
    res = await fetch(cfg.tokenUrl, { method: "POST", headers: { "content-type": contentType }, body });
  } catch (err) {
    throw new Error(`refresh request to ${cfg.tokenUrl} failed: ${err instanceof Error ? err.message : String(err)}`);
  }
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`refresh for "${name}" returned ${res.status}: ${text.slice(0, 200)}`);
  }
  let json: unknown;
  try {
    json = await res.json();
  } catch {
    throw new Error(`refresh for "${name}" returned non-JSON`);
  }
  return mapTokenResponse(cfg, json);
}

File: gateway/src/providers.rs
//! Owned OAuth refresh (ports `src/providers.ts`). Per-provider public
//! constants + a standard `refresh_token` grant; no dependency on pi's code.
//! See docs/rust-port.md (P3).

use crate::proxy::BErr;
use crate::util::now_ms;
use serde_json::Value;

#[derive(Clone, Copy)]
pub enum Encoding {
    Form,
    Json,
}

pub struct ProviderConfig {
    pub token_url: &'static str,
    pub client_id: &'static str,
    pub encoding: Encoding,
    pub skew_ms: i64,
}

pub fn provider(name: &str) -> Option<ProviderConfig> {
    match name {
        "openai-codex" => Some(ProviderConfig {
            token_url: "https://auth.openai.com/oauth/token",
            client_id: "app_EMoamEEZ73f0CkXaXp7hrann",
            encoding: Encoding::Form,
            skew_ms: 0,
        }),
        "anthropic" => Some(ProviderConfig {
            token_url: "https://platform.claude.com/v1/oauth/token",
            client_id: "9d1c250a-e61b-44d9-88ed-5944d1962f5e",
            encoding: Encoding::Json,
            skew_ms: 5 * 60 * 1000,
        }),
        _ => None,
    }
}

pub struct RefreshedTokens {
    pub access: String,
    pub refresh: String,
    pub expires: i64,
}

/// Map a provider token-endpoint response to our credential fields. Pure aside
/// from `now_ms`. Errors on a malformed response so the caller fails closed.
pub fn map_token_response(cfg: &ProviderConfig, json: &Value) -> Result<RefreshedTokens, BErr> {
    let access = json.get("access_token").and_then(|v| v.as_str());
    let refresh = json.get("refresh_token").and_then(|v| v.as_str());
    let expires_in = json.get("expires_in").and_then(|v| v.as_f64());
    match (access, refresh, expires_in) {
        (Some(a), Some(r), Some(e)) => Ok(RefreshedTokens {
            access: a.to_string(),
            refresh: r.to_string(),
            expires: now_ms() + (e as i64) * 1000 - cfg.skew_ms,
        }),
        _ => Err(format!("token response missing fields: {json}").into()),
    }
}

/// Perform the `refresh_token` grant. A direct host-side call (the gateway's own
/// trusted action). Throws on any failure — the caller must fail closed.
pub async fn refresh(name: &str, refresh_token: &str) -> Result<RefreshedTokens, BErr> {
    let cfg = provider(name).ok_or_else(|| format!("no refresh config for provider \"{name}\""))?;
    let client = reqwest::Client::new();
    let req = match cfg.encoding {
        Encoding::Form => client.post(cfg.token_url).form(&[
            ("grant_type", "refresh_token"),
            ("refresh_token", refresh_token),
            ("client_id", cfg.client_id),
        ]),
        Encoding::Json => client.post(cfg.token_url).json(&serde_json::json!({
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
            "client_id": cfg.client_id,
        })),
    };
    let resp = req
        .send()
        .await
        .map_err(|e| format!("refresh request to {} failed: {e}", cfg.token_url))?;
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if !status.is_success() {
        let snippet = &text[..text.len().min(200)];
        return Err(format!("refresh for \"{name}\" returned {status}: {snippet}").into());
    }
    let json: Value = serde_json::from_str(&text).map_err(|_| format!("refresh for \"{name}\" returned non-JSON"))?;
    map_token_response(&cfg, &json)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_fields_and_applies_skew() {
        let cfg = provider("openai-codex").unwrap();
        let before = now_ms();
        let out = map_token_response(
            &cfg,
            &serde_json::json!({"access_token":"a","refresh_token":"r","expires_in":3600}),
        )
        .unwrap();
        assert_eq!(out.access, "a");
        assert_eq!(out.refresh, "r");
        assert!(out.expires >= before + 3_600_000 && out.expires <= now_ms() + 3_600_000);
    }

    #[test]
    fn anthropic_subtracts_its_skew() {
        let cfg = provider("anthropic").unwrap();
        let out = map_token_response(
            &cfg,
            &serde_json::json!({"access_token":"a","refresh_token":"r","expires_in":3600}),
        )
        .unwrap();
        assert!(out.expires <= now_ms() + 3_600_000 - 5 * 60 * 1000);
    }

    #[test]
    fn malformed_response_fails_closed() {
        let cfg = provider("openai-codex").unwrap();
        assert!(map_token_response(&cfg, &serde_json::json!({"access_token":"a"})).is_err());
        assert!(map_token_response(&cfg, &serde_json::json!({"error":"invalid_grant"})).is_err());
    }
}
